python/tchannel/tornado/connection.py

```diff
@@ -294,8 +293,15 @@ def write(self, message):
             message_factory = self.response_message_factory
 
         fragments = message_factory.fragment(message)
+
+        write_future = self._write(fragments.next())
+
         for fragment in fragments:
-            yield self._write(fragment)
+            write_future.add_done_callback(
+                lambda f: f.exception() or self.write(fragment)
```

Comment:
Yeah, this is wrong.

You need to create a result future that you return and if `f.exception()`, you need to `set_exception` (or better, `set_exc_info`) on it. Otherwise, you do `set_result(None)` on it after all chunks are done writing. Currently there's no write-order guarantee either.

Oh and it does weird things with the function scope.

```
>>> def go():
...     functions = []
...     for i in range(5):
...         functions.append(lambda: print(i))
...     return functions
...
>>> for f in go():
...     f()
...
4
4
4
4
4
```